src/aws_encryption_sdk/keyrings/aws_kms/__init__.py

```diff
@@ -180,25 +181,26 @@ class _AwsKmsSingleCmkKeyring(Keyring):
     def on_encrypt(self, encryption_materials):
         # type: (EncryptionMaterials) -> EncryptionMaterials
         trace_info = MasterKeyInfo(provider_id=_PROVIDER_ID, key_info=self._key_id)
+        new_materials = copy.copy(encryption_materials)
```

Comment:
It's not strictly necessary, but I think that having access to the original materials inside that scope could help with debugging if things aren't doing what people thing they should.

Reply:
Maybe this is a Python subtlety I'm missing, but aren't the original materials already available through the "encryption_materials" input?

Reply:
The flow is:

```
if no data key:
    data key, EDK = generate()
    new materials = materials(with data key)
else:
    EDK = encrypt(data key)

new materials = materials with data key(with EDK)
```

I went with that flow so that we could single-source the "with EDK" bit, but because of that, if we want to retain an available reference to the materials in the request we have a choice: we can either copy the request materials first:

```
copied materials = copy(request materials)

if no data key:
    data key, EDK = generate()
    copied materials = copied materials(with data key)
else:
    EDK = encrypt(data key)

copied materials = copied materials(with EDK)
```

or we can balance a null value

```
new materials = None

if no data key:
    data key, EDK = generate()
    new materials = request materials(with data key)
else:
    EDK = encrypt(data key)

if new materials is None:
    new materials = request materials(with EDK)
else:
    new materials = new materials(with EDK)
```

The copy flow seems cleaner to me.

Reply:
This doesn't work?

```python
new materials = request materials

if no data key:
    data key, EDK = generate()
    new materials = new materials(with data key)
else:
    EDK = encrypt(data key)

new materials = new materials(with EDK)
```

Reply:
Yeah, I guess that would be fine. It feels less explicit, but since we're only ever using the immutable APIs it would effectively be doing the same thing.

Reply:
That's how I did it in Java, but I'll defer to your judgement on what you prefer for Python